services/signer/standard/helpers.go

```diff
@@ -165,3 +172,54 @@ func (s *Service) signRootsByAccountType(ctx context.Context, accounts []e2wtype
 	}
 	return sigs, nil
 }
+
+// deduplicateAccountRootPairs deduplicates (account, root) pairs to avoid duplicate signing requests.
+// It returns unique accounts and data along with a mapping from original indices to unique indices.
+func deduplicateAccountRootPairs(accounts []e2wtypes.Account, data [][]byte) deduplicationResult {
+	type accountRootPair struct {
+		accountKey string
+		rootKey    string
+	}
+
+	// Map to first occurrence index.
+	uniquePairs := make(map[accountRootPair]int)
+	var uniqueAccounts []e2wtypes.Account
+	var uniqueData [][]byte
+	originalToUniqueIndex := make([]int, len(accounts))
+
+	for i := range accounts {
+		accountKey := accounts[i].ID().String()
+		rootKey := string(data[i])
+		pair := accountRootPair{accountKey: accountKey, rootKey: rootKey}
+
+		if uniqueIndex, exists := uniquePairs[pair]; exists {
+			originalToUniqueIndex[i] = uniqueIndex
+		} else {
+			uniqueIndex = len(uniqueAccounts)
+			uniquePairs[pair] = uniqueIndex
+			originalToUniqueIndex[i] = uniqueIndex
+			uniqueAccounts = append(uniqueAccounts, accounts[i])
+			uniqueData = append(uniqueData, data[i])
+		}
+	}
+
+	return deduplicationResult{
+		uniqueAccounts:        uniqueAccounts,
+		uniqueData:            uniqueData,
+		originalToUniqueIndex: originalToUniqueIndex,
+	}
+}
+
+// mapSignaturesToOriginalPositions maps unique signatures back to the original request positions.
+func mapSignaturesToOriginalPositions(
+	sigs []phase0.BLSSignature,
+	uniqueSigs []e2types.Signature,
+	dedup deduplicationResult,
+) {
+	for i := range sigs {
+		uniqueIndex := dedup.originalToUniqueIndex[i]
+		if uniqueIndex < len(uniqueSigs) && uniqueSigs[uniqueIndex] != nil {
```

Comment:
If any of these conditions are not met we will silently ignore that signature. Is there a reason for this? If not I think we should return an appropriate error.